Add UT for route agent health checker

aswinsuryan opened the pull request:
Add UT for route agent health checker
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

aswinsuryan committed:
Add UT for route agent health checker

Signed-off-by: Aswin Suryanarayanan <asuryana@redhat.com>

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/health_checker_suite_test.go:
```diff
@@ -0,0 +1,43 @@
+
+func TestHealthChecker(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "RouteAgent Health Checker")
```
```suggestion
	RunSpecs(t, "RouteAgent Health Checker Suite")
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+		t.stopCh = make(chan struct{})
+		scheme := runtime.NewScheme()
+		Expect(submarinerv1.AddToScheme(scheme)).To(Succeed())
+		Expect(submarinerv1.AddToScheme(kubeScheme.Scheme)).To(Succeed())
```
This is already done in `BeforeSuite`.

```suggestion
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+	BeforeEach(func() {
+		t.stopCh = make(chan struct{})
+		scheme := runtime.NewScheme()
+		Expect(submarinerv1.AddToScheme(scheme)).To(Succeed())
```
This isn't needed - `scheme` is unused.

```suggestion
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+			healthCheckIP1: fake.NewPinger(healthCheckIP1),
+		}
+
+		t.localNodeName = "nodeName"
```
This is really a constant and can be defined as such.

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+			PingInterval:             1, // Set interval to 1 second for faster testing
+			MaxPacketLossCount:       1,
+			HealthCheckerEnabled:     true,
+			RouteAgentUpdateInterval: 2 * time.Second,
```
You can make this a lot less to speed up the tests

```suggestion
			RouteAgentUpdateInterval: 100 * time.Millisecond,
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+		}
+
+		t.handler = healthchecker.New(config, t.client, "v1", t.localNodeName)
+		Expect(t.handler.Init()).To(Succeed())
```
You shouldn't call `Init` - ControllerSupport does that.

```suggestion
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,277 @@
+	When("syncRouteAgentStatus is triggered", func() {
+		It("should create or update the RouteAgent object and check LatencyInfo", func() {
+			endpoint1 := t.CreateEndpoint(t.newSubmEndpoint())
+			err = t.handler.RemoteEndpointCreated(endpoint1)
```
You shouldn't call `RemoteEndpointCreated` directly - the ControllerSupport will do that when triggered by `Endpoint` creation.

aswinsuryan replied:
Done.

tpantelis commented:
I think we should structure the test specs as follows:

```
var _ = Describe("RouteAgent syncing", func() {
    It("should create a RouteAgent resource", func() {
         // await the RouteAgent - no RemoteEndpoint info expected
    })

    When("a remote Endpoint is created", func() {
        It("should add its RemoteEndpoint information to the RouteAgent resource", func() {
            // await the RemoteEndpoint and verify the Spec
        })
    })

    When("a remote Endpoint is updated", func() {
        It("should update its RemoteEndpoint information in the RouteAgent resource", ...)

    When("a remote Endpoint is deleted", func() {
        It("should remove its RemoteEndpoint information from the RouteAgent resource", ...)
})

var _ = Describe("RemoteEndpoint latency info", func() {
    When("a remote Endpoint is created", func() {
        It("should start a pinger and correctly update the RemoteEndpoint Status and LatencyInfo", func() {
            // await the RemoteEndpoint and verify the LatencyInfo and Status
        })

        Context("with no HealthCheckIP", func() {
		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
	})

        Context("with health check not enabled", func() {
		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
	})

        When("on the gateway", func() {
		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
                 // Call ControllerSupport.CreateLocalHostEndpoint() before creating the remote EP
	})
    })

    When("a remote Endpoint is updated", func() {
        Context("and the HealthCheckIP was changed", func() {
            It("should stop the pinger and start a new one", ...)
        })

        Context("and the HealthCheckIP did not change", func() {
            It("should not start a new pinger", ...)
        })
    })

    When("a remote Endpoint is deleted", func() {
        It("should stop the pinger", ...)
})
```

aswinsuryan commented:
> I think we should structure the test specs as follows:
> 
> ```
> var _ = Describe("RouteAgent syncing", func() {
>     It("should create a RouteAgent resource", func() {
>          // await the RouteAgent - no RemoteEndpoint info expected
>     })
> 
>     When("a remote Endpoint is created", func() {
>         It("should add its RemoteEndpoint information to the RouteAgent resource", func() {
>             // await the RemoteEndpoint and verify the Spec
>         })
>     })
> 
>     When("a remote Endpoint is updated", func() {
>         It("should update its RemoteEndpoint information in the RouteAgent resource", ...)
> 
>     When("a remote Endpoint is deleted", func() {
>         It("should remove its RemoteEndpoint information from the RouteAgent resource", ...)
> })
> 
> var _ = Describe("RemoteEndpoint latency info", func() {
>     When("a remote Endpoint is created", func() {
>         It("should start a pinger and correctly update the RemoteEndpoint Status and LatencyInfo", func() {
>             // await the RemoteEndpoint and verify the LatencyInfo and Status
>         })
> 
>         Context("with no HealthCheckIP", func() {
> 		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
> 	})
> 
>         Context("with health check not enabled", func() {
> 		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
> 	})
> 
>         When("on the gateway", func() {
> 		It("should not start a pinger and should set the RemoteEndpoint Status to None", ...)
>                  // Call ControllerSupport.CreateLocalHostEndpoint() before creating the remote EP
> 	})
>     })
> 
>     When("a remote Endpoint is updated", func() {
>         Context("and the HealthCheckIP was changed", func() {
>             It("should stop the pinger and start a new one", ...)
>         })
> 
>         Context("and the HealthCheckIP did not change", func() {
>             It("should not start a new pinger", ...)
>         })
>     })
> 
>     When("a remote Endpoint is deleted", func() {
>         It("should stop the pinger", ...)
> })
> ```

Thanks for the suggestion . I updated the structure .

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+		var routeAgent *submarinerv1.RouteAgent
+		Eventually(func() bool {
+			var err error
+			routeAgent, err = t.client.Get(context.TODO(), localNodeName, metav1.GetOptions{})
+			return err == nil && routeAgent != nil
```
We should fail on unexpected error.

```suggestion
			routeAgent, err = t.client.Get(context.TODO(), localNodeName, metav1.GetOptions{})
			if apierrors.IsNotFound(err) {
			    return false
			}
			
			Expect(err).To(Succeed())
			return true
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+			var err error
+			routeAgent, err = t.client.Get(context.TODO(), localNodeName, metav1.GetOptions{})
+			return err == nil && routeAgent != nil
+		}, time.Second*50, time.Millisecond*500).Should(BeTrue())
```
50 s is way too long as is 500ms interval - the default  is 100ms max - that should be plenty.

```suggestion
		}).Should(BeTrue())
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+			routeAgent, err = t.client.Get(context.TODO(), localNodeName, metav1.GetOptions{})
+			return err == nil && routeAgent != nil
+		}, time.Second*50, time.Millisecond*500).Should(BeTrue())
+		Expect(routeAgent.Name).To(Equal(localNodeName))
+		Expect(routeAgent.Namespace).To(Equal(namespace))
```
No need to verify these since they're used in the `Get` query.

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+	When("a remote Endpoint is updated", func() {
+		It("should update its RemoteEndpoint information in the RouteAgent resource", func() {
+			endpoint1 := t.CreateEndpoint(t.newSubmEndpoint(healthCheckIP1))
+			t.pingerMap[healthCheckIP1].AwaitStart()
+			endpoint1.Spec.Hostname = "newHostName"
+			t.UpdateEndpoint(endpoint1)
+
+			pingerObject, found := t.pingerMap[endpoint1.Spec.HealthCheckIP]
+			Expect(found).To(BeTrue())
+			Expect(pingerObject.GetIP()).To(Equal(healthCheckIP1))
```
This test is to just verify the remote endpoint spec. The pinger stuff is verified in another spec.

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+			Expect(found).To(BeTrue())
+			Expect(pingerObject.GetIP()).To(Equal(healthCheckIP1))
+
+			var routeAgent *submarinerv1.RouteAgent
```
You should first verify the original EP spec. You really want to do the same as in L64 above but instead of duplicating code, I suggest to combine the two. So L64+ would become:

```
When("a remote Endpoint is created/updated", func() {
    It("should add/update its RemoteEndpoint information to the RouteAgent resource", func() {
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+
+	When("a remote Endpoint is deleted", func() {
+		It("should remove its RemoteEndpoint information from the RouteAgent resource", func() {
+			endpoint1 := t.CreateEndpoint(t.newSubmEndpoint(healthCheckIP1))
```
Similarly as with update, you can combine delete as well:

`When("a remote Endpoint is created/updated/deleted", func() {`

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+		t := newTestDriver(true)
+		It("should start a pinger and correctly update the RemoteEndpoint Status and LatencyInfo", func() {
+			endpoint1 := t.CreateEndpoint(t.newSubmEndpoint(healthCheckIP1))
+
```
First verify the pinger was started.

```suggestion
t.pingerMap[healthCheckIP1].AwaitStart()
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+		Context("with no HealthCheckIP", func() {
+			It("should not start a pinger and should set the RemoteEndpoint Status to None", func() {
+				endpoint1 := t.CreateEndpoint(t.newSubmEndpoint(""))
+				var routeAgent *submarinerv1.RouteAgent
```
```suggestion
                                 t.pingerMap[healthCheckIP1].AwaitNoStart()
                                 
				var routeAgent *submarinerv1.RouteAgent
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+			It("should not start a pinger and should set the RemoteEndpoint Status to None", func() {
+				_ = t.CreateLocalHostEndpoint()
+				endpoint1 := t.CreateEndpoint(t.newSubmEndpoint(healthCheckIP1))
+
```
```suggestion
t.pingerMap[healthCheckIP1].AwaitNoStart()
```

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+
+	When("a remote Endpoint is created", func() {
+		t := newTestDriver(false)
+		Context("with health check not enabled", func() {
```
This should be within L144.

aswinsuryan replied:
Done.

tpantelis commented on pkg/routeagent_driver/handlers/healthchecker/healthchecker_test.go:
```diff
@@ -0,0 +1,409 @@
+			Expect(ok).To(BeTrue())
+
+			return p
+		}
```
Instead of passing in `enableHealthCheck`  to `newTestDriver`, it's better to default  `config.HealthCheckerEnabled` to true and have spec(s) set it to false in their own `BeforeEach` and move L357-359 below to `JustBeforeEach`.

aswinsuryan replied:
Done.

aswinsuryan commented:
@tpantelis there is a race condition being reported in the set state method, where do you think we need to fix it ? some of the  test infra ? Or do you think it is the code itself?

WARNING: DATA RACE
Read at 0x00c0000e3860 by goroutine 21:
  github.com/submariner-io/submariner/pkg/event.(*HandlerBase).State()
      /go/src/github.com/submariner-io/submariner/pkg/event/handler.go:116 +0x30c
  github.com/submariner-io/submariner/pkg/routeagent_driver/handlers/healthchecker.(*controller).syncRouteAgentStatus()

tpantelis committed:
Make HandlerBase.handlerState field atomic

...to avoid a race condition if State is called before SetState.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
> @tpantelis there is a race condition being reported in the set state method, where do you think we need to fix it ? some of the test infra ? Or do you think it is the code itself?

I pushed https://github.com/submariner-io/submariner/pull/3152/commits/8a30449f7e30afee47b055859e8fcdcb46698619 to address it.

tpantelis committed:
Adustments to route agent health checker unit tests

- Refactor code to await the RouteAgent resource to funntions
  to avoid duplication

- Add spec to test pinger connection error

- Other minor changes

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
I had additional minor adjustments that I just pushed in [another commit](https://github.com/submariner-io/submariner/pull/3152/commits/6198d81f408e55f18f2b162c3c29a8747cd363e7) rather then adding comments (thought it would be easier).
